ecs-agent/tmds/handlers/fault/v1/handlers/handlers.go

```diff
@@ -692,3 +736,131 @@ func validateTaskNetworkConfig(taskNetworkConfig *state.TaskNetworkConfig) error
+		}
+	}
+
+	return nil
+}
+
+// checkTCFault check if there's existing network-latency fault or network-packet-loss fault.
+func (h *FaultHandler) checkTCFault(taskMetadata *state.TaskResponse) (bool, bool, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	interfaceName := taskMetadata.TaskNetworkConfig.NetworkNamespaces[0].NetworkInterfaces[0].DeviceName
+	networkMode := taskMetadata.TaskNetworkConfig.NetworkMode
+	// If task's network mode is awsvpc, we need to run nsenter to access the task's network namespace.
+	nsenterPrefix := ""
+	if networkMode == ecs.NetworkModeAwsvpc {
+		nsenterPrefix = fmt.Sprintf(nsenterCommandString, taskMetadata.TaskNetworkConfig.NetworkNamespaces[0].Path)
+	}
+
+	// We will run the following Linux command to assess if there existing fault.
+	// "tc -j q show dev {INTERFACE} parent 1:1"
+	// The command above gives the output of "tc q show dev {INTERFACE} parent 1:1" in json format.
+	// We will then unmarshall the json string and evaluate the fields of it.
+	tcCheckInjectionCommandComposed := nsenterPrefix + fmt.Sprintf(tcCheckInjectionCommandString, interfaceName)
+	cmdOutput, err := h.runExecCommand(ctx, tcCheckInjectionCommandComposed)
+	if err != nil {
+		return false, false, errors.New("failed to check existing network fault: " + err.Error())
+	}
+	// Log the command output to better help us debug.
+	logger.Info(fmt.Sprintf("'%s' command result: '%s'", tcCheckInjectionCommandComposed, string(cmdOutput[:])))
+
+	// Check whether latency fault exists and whether packet loss fault exists separately.
+	latencyFaultExists, err := checkLatencyFault(cmdOutput)
+	if err != nil {
+		return false, false, errors.New("failed to check existing network fault: " + err.Error())
+	}
+	packetLossFaultExists, err := checkPacketLossFault(cmdOutput)
+	if err != nil {
+		return false, false, errors.New("failed to check existing network fault: " + err.Error())
+	}
+	return latencyFaultExists, packetLossFaultExists, nil
+}
+
+// checkLatencyFault parses the tc command output and checks if there's existing network-latency fault running.
+func checkLatencyFault(commandOutputInBytes []byte) (bool, error) {
+	var outputUnmarshalled []map[string]interface{}
+	err := json.Unmarshal(commandOutputInBytes, &outputUnmarshalled)
+	if err != nil {
+		return false, errors.New("failed to unmarshal tc command output: " + err.Error())
+	}
+	for _, line := range outputUnmarshalled {
+		// Check if field "kind":"netem" exists.
+		if line["kind"] == "netem" {
+			// Now check if network packet loss fault exists.
+			if options := line["options"]; options != nil {
+				if delay := options.(map[string]interface{})["delay"]; delay != nil {
+					return true, nil
+				}
+			}
+		}
+	}
+	return false, nil
+}
+
+// checkPacketLossFault parses the tc command output and checks if there's existing network-packet-loss fault running.
+func checkPacketLossFault(commandOutputInBytes []byte) (bool, error) {
```

Comment:
non-blocking - this one looks similar to the one above. probably we can re-use one of them by refactoring it a bit.

Reply:
Yeah we can re-use the unmarshal part and a part of the for loop